File: frontend/types/communication.ts
export type AnnouncementAudience = 'all' | 'admin' | 'teacher';

export interface Announcement {
  id: string;
  title: string;
  body: string;
  target_audience: AnnouncementAudience;
  sent_by_name: string;
  created_at: string;
}

export interface CreateAnnouncementRequest {
  title: string;
  body: string;
  target_audience: AnnouncementAudience;
}

export interface MessageResponse {
  id: string;
  sender_id: string;
  sender_name: string;
  recipient_id: string | null;
  recipient_name: string;
  recipients: MessageRecipient[];
  subject: string | null;
  body: string;
  is_read: boolean;
  thread_id: string;
  created_at: string;
}

export interface SendMessageRequest {
  recipient_id?: string;
  recipient_ids?: string[];
  recipient_group?: 'all_teachers' | 'all_admins';
  body: string;
  subject?: string;
  thread_id?: string;
  parent_message_id?: string;
}

export interface MessageRecipient {
  id: string;
  name: string;
  role: 'super_admin' | 'admin' | 'teacher';
}

export interface RecipientPage {
  total: number;
  page: number;
  per_page: number;
  items: MessageRecipient[];
}

export interface ThreadResponse {
  thread_id: string;
  items: MessageResponse[];
}

export interface InboxResponse {
  total: number;
  page: number;
  per_page: number;
  items: MessageResponse[];
  unread_count: number;
}

export interface UnreadCountResponse {
  count: number;
}

File: backend/app/schemas/communication.py
import uuid
from datetime import datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


class TargetAudienceEnum(str, Enum):
    all = "all"
    admin = "admin"
    teacher = "teacher"


# ---------------------------------------------------------------------------
# Announcements
# ---------------------------------------------------------------------------

class AnnouncementCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=255)
    body: str = Field(..., min_length=1)
    target_audience: TargetAudienceEnum


class AnnouncementResponse(BaseModel):
    id: uuid.UUID
    title: str
    body: str
    target_audience: TargetAudienceEnum
    sent_by_name: str
    created_at: datetime

    model_config = {"from_attributes": True}


class PaginatedAnnouncementResponse(BaseModel):
    total: int
    page: int
    per_page: int
    items: list[AnnouncementResponse]


# ---------------------------------------------------------------------------
# Messages
# ---------------------------------------------------------------------------

class SendMessageRequest(BaseModel):
    recipient_id: Optional[uuid.UUID] = None
    recipient_ids: list[uuid.UUID] = Field(default_factory=list, max_length=100)
    recipient_group: Optional[str] = Field(None, pattern="^(all_teachers|all_admins)$")
    subject: Optional[str] = Field(None, max_length=255)
    body: str = Field(..., min_length=1)
    thread_id: Optional[uuid.UUID] = None
    parent_message_id: Optional[uuid.UUID] = None


class MessageRecipientResponse(BaseModel):
    id: uuid.UUID
    name: str
    role: str


class MessageResponse(BaseModel):
    id: uuid.UUID
    sender_id: uuid.UUID
    sender_name: str
    recipient_id: Optional[uuid.UUID]
    recipient_name: str
    recipients: list[MessageRecipientResponse]
    subject: Optional[str]
    body: str
    is_read: bool
    thread_id: uuid.UUID
    created_at: datetime

    model_config = {"from_attributes": True}


class PaginatedMessageResponse(BaseModel):
    total: int
    page: int
    per_page: int
    items: list[MessageResponse]


class InboxResponse(BaseModel):
    total: int
    page: int
    per_page: int
    items: list[MessageResponse]
    unread_count: int


class UnreadCountResponse(BaseModel):
    count: int


class RecipientResponse(BaseModel):
    id: uuid.UUID
    name: str
    role: str


class PaginatedRecipientResponse(BaseModel):
    total: int
    page: int
    per_page: int
    items: list[RecipientResponse]


class ThreadResponse(BaseModel):
    thread_id: uuid.UUID
    items: list[MessageResponse]
